federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	return client.AutoscalingV1().HorizontalPodAutoscalers(hpa.Namespace).Update(hpa)
+}
+
+func (a *HpaAdapter) ClusterWatch(client kubeclientset.Interface, namespace string, options metav1.ListOptions) (watch.Interface, error) {
+	return client.AutoscalingV1().HorizontalPodAutoscalers(namespace).Watch(options)
+}
+
+func (a *HpaAdapter) NewTestObject(namespace string) pkgruntime.Object {
+	var min int32 = 4
+	var targetCPU int32 = 70
+	return &autoscalingv1.HorizontalPodAutoscaler{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "test-hpa-",
+			Namespace:    namespace,
+		},
+		Spec: autoscalingv1.HorizontalPodAutoscalerSpec{
+			ScaleTargetRef: autoscalingv1.CrossVersionObjectReference{
+				Kind: "replicaset",
+				Name: "myrs",
+			},
+			MinReplicas:                    &min,
+			MaxReplicas:                    int32(10),
+			TargetCPUUtilizationPercentage: &targetCPU,
+		},
+	}
+}
+
+func (a *HpaAdapter) ImplementsReconcilePlugin() bool {
+	return true
+}
+
+type replicaNums struct {
+	min int32
+	max int32
+}
+
+type hpaLists struct {
+	availableMin sets.String
+	availableMax sets.String
+	noHpa        sets.String
+}
+
+func (a *HpaAdapter) ReconcileHook(fedObj pkgruntime.Object, currentObjs map[string]pkgruntime.Object) (map[string]pkgruntime.Object, error) {
+	fedHpa := fedObj.(*autoscalingv1.HorizontalPodAutoscaler)
+	requestedMin := int32(1)
+	if fedHpa.Spec.MinReplicas != nil {
+		requestedMin = *fedHpa.Spec.MinReplicas
+	}
+	requestedReplicas := replicaNums{
+		min: requestedMin,
+		max: fedHpa.Spec.MaxReplicas,
+	}
+	// replica distribution count, per cluster
+	rdc := replicaNums{
+		min: requestedReplicas.min / int32(len(currentObjs)),
+		max: requestedReplicas.max / int32(len(currentObjs)),
+	}
+	if rdc.min < 1 {
+		rdc.min = 1
+	}
+	// TODO: Is there a better way?
+	// We need to cap the lowest limit of Max to 2, because in a
+	// situation like both min and max become 1 (same) for all clusters,
+	// no rebalancing would happen.
+	if rdc.max < 2 {
+		rdc.max = 2
+	}
+
+	// Prepare pass: Analyse existing local hpa's if any
+	// clusterLists holds the list of those clusters which can offer
+	// min and max replicas, to those which want them.
+	// For example new clusters joining the federation and/or
+	// those clusters which need to increase or reduce replicas
+	// beyond min/max limits.
+	// schedStatus currently have status of existing hpas.
+	// It will eventually have desired status for this reconcile.
+	clusterLists, currentReplicas, scheduleStatus := prepareForScheduling(currentObjs)
+
+	remainingReplicas := replicaNums{
+		min: requestedReplicas.min - currentReplicas.min,
+		max: requestedReplicas.max - currentReplicas.max,
+	}
+
+	// Pass 1: reduction of replicas if needed ( situation that fedHpa updated replicas
+	// to lesser then existing).
+	// In this pass, we remain pessimistic and reduce one replica per cluster at a time.
```

Comment:
I thought of keeping this to a pessimistic lowest, until the "slowness of using this value" could be established. I did it to avoid any thrashing whatsoever, using an aggressive value. As you mentioned can be done later.